core/async/src/test_loop/event_handler.rs

```diff
@@ -0,0 +1,135 @@
+/// handles only some events, so we will usually have multiple event handlers
+/// registered to cover all event types.
+pub trait LoopEventHandler<Data, Event> {
+    /// Called once, when the loop runs for the first time.
+    fn init(&mut self, _sender: DelaySender<Event>) {}
+
+    /// Handles an event. If this handler indeed handles the event, it should
+    /// return None after handling it. Otherwise, it should return Some with
+    /// the same event that was passed in, so that it can be given to the next
+    /// event handler.
+    fn handle(&mut self, event: Event, data: &mut Data) -> Option<Event>;
+}
+
+/// Functions that can be called on LoopEventHandler to adapt it to different
+/// situations.
+pub trait LoopEventHandlerHelpers<Data, Event> {
+    /// Adapts this handler to a handler whose data is a superset of our data
+    /// and whose event is a superset of our event.
+    ///   For data, A is a superset of B if A implements AsRef<B> and AsMut<B>.
+    ///   For event, A is a superset of B if A implements From<B> and
+    ///     TryIntoOrSelf<B>.
+    fn widen(self) -> WideningEventHandler<Data, Event>;
+
+    /// Adapts this handler to a handler whose data is a vector of our data,
+    /// and whose event is a is the tuple (index, our event), for a specific
+    /// index.
+    fn for_index(self, index: usize) -> IndexedLoopEventHandler<Data, Event>;
+}
+
+impl<Data, Event, T: LoopEventHandler<Data, Event> + Sized + 'static>
+    LoopEventHandlerHelpers<Data, Event> for T
+{
+    fn widen(self) -> WideningEventHandler<Data, Event> {
+        WideningEventHandler { inner: Box::new(self) }
+    }
+
+    fn for_index(self, index: usize) -> IndexedLoopEventHandler<Data, Event> {
+        IndexedLoopEventHandler::new(Box::new(self), index)
+    }
+}
+
+/// A convenient trait to TryInto, or else return the original object. It's useful
+/// for implementing event handlers.
+pub trait TryIntoOrSelf<R>: Sized {
+    fn try_into_or_self(self) -> Result<R, Self>;
+}
+
+impl<R, T: TryInto<R, Error = T>> TryIntoOrSelf<R> for T {
+    fn try_into_or_self(self) -> Result<R, Self> {
+        self.try_into()
+    }
+}
+
+/// Implements .widen() for an event handler.
+pub struct WideningEventHandler<Data, Event> {
+    inner: Box<dyn LoopEventHandler<Data, Event>>,
+}
+
+impl<
+        Data,
+        Event,
+        OuterData: AsMut<Data>,
+        OuterEvent: TryIntoOrSelf<Event> + From<Event> + 'static,
+    > LoopEventHandler<OuterData, OuterEvent> for WideningEventHandler<Data, Event>
+{
+    fn init(&mut self, sender: DelaySender<OuterEvent>) {
+        self.inner.init(sender.narrow())
+    }
+
+    fn handle(&mut self, event: OuterEvent, data: &mut OuterData) -> Option<OuterEvent> {
+        let mut inner_data = data.as_mut();
+        match event.try_into_or_self() {
+            Ok(e) => self.inner.handle(e, &mut inner_data).map(|e| e.into()),
+            Err(e) => Some(e),
```

Comment:
naming both event and error as "e" in a single context poorly affects readability. How about full names, or at least "ev" and "err"?